File: platform/macos/backend.c
#include "wlf/platform/macos/backend.h"
#include "wlf/types/wlf_output.h"
#include "wlf/utils/wlf_linked_list.h"
#include "wlf/utils/wlf_signal.h"
#include "wlf/utils/wlf_log.h"

#include <stdlib.h>
#include <assert.h>
#include <string.h>

/**
 * @brief macOS output manager specific data
 */
struct wlf_macos_output_manager {
	struct wlf_output_manager base;  /**< Base output manager structure */
};

static void macos_output_manager_destroy_impl(struct wlf_output_manager *manager) {
	struct wlf_macos_output_manager *macos_manager =
		(struct wlf_macos_output_manager *)manager;
	free(macos_manager);
}

static const struct wlf_output_manager_impl macos_output_manager_impl = {
	.destroy = macos_output_manager_destroy_impl,
};

static struct wlf_output_manager *macos_output_manager_create(void) {
	struct wlf_macos_output_manager *manager = calloc(1, sizeof(struct wlf_macos_output_manager));
	if (manager == NULL) {
		wlf_log_errno(WLF_ERROR, "Failed to allocate macOS output manager");
		return NULL;
	}

	wlf_output_manager_init(&manager->base, &macos_output_manager_impl);
	wlf_log(WLF_DEBUG, "Created macOS output manager");

	return &manager->base;
}

static void macos_output_manager_destroy(struct wlf_listener *listener, void *data) {
	struct wlf_backend_macos *backend =
		wlf_container_of(listener, backend, listeners.output_manager_destroy);
	wlf_output_manager_destroy(backend->base.output_manager);
	backend->base.output_manager = NULL;
}

static bool macos_backend_start(struct wlf_backend *backend) {
	struct wlf_backend_macos *macos = (struct wlf_backend_macos *)backend;

	if (macos->started) {
		wlf_log(WLF_INFO, "macOS backend already started");
		return true;
	}

	// Create output manager for macOS
	backend->output_manager = macos_output_manager_create();
	if (backend->output_manager == NULL) {
		wlf_log(WLF_ERROR, "Failed to create output manager for macOS backend");
		return false;
	}

	macos->listeners.output_manager_destroy.notify = macos_output_manager_destroy;
	wlf_signal_add(&backend->output_manager->events.destroy, &macos->listeners.output_manager_destroy);

	// TODO: Enumerate displays and create outputs
	// This would involve using NSScreen or CGDisplay APIs

	macos->started = true;
	wlf_log(WLF_INFO, "macOS backend started successfully");

	return true;
}

static void macos_backend_stop(struct wlf_backend *backend) {
	struct wlf_backend_macos *macos = (struct wlf_backend_macos *)backend;

	if (!macos->started) {
		return;
	}

	wlf_log(WLF_DEBUG, "Stopping macOS backend");

	if (backend->output_manager) {
		wlf_output_manager_destroy(backend->output_manager);
		backend->output_manager = NULL;
	}

	macos->started = false;
}

static void macos_backend_destroy(struct wlf_backend *backend) {
	struct wlf_backend_macos *macos = (struct wlf_backend_macos *)backend;
	wlf_log(WLF_DEBUG, "Destroying macOS backend");

	macos_backend_stop(backend);
	wlf_linked_list_remove(&macos->listeners.output_manager_destroy.link);

	free(macos);
}

static const struct wlf_backend_impl macos_backend_impl = {
	.start = macos_backend_start,
	.stop = macos_backend_stop,
	.destroy = macos_backend_destroy,
};

static struct wlf_backend *macos_backend_create(void *args) {
	struct wlf_backend_macos *backend = calloc(1, sizeof(struct wlf_backend_macos));
	if (backend == NULL) {
		wlf_log_errno(WLF_ERROR, "Failed to allocate macOS backend");
		return NULL;
	}

	backend->base.impl = &macos_backend_impl;
	backend->base.type = WLF_BACKEND_MACOS;
	backend->base.data = backend;
	wlf_signal_init(&backend->base.events.destroy);

	backend->started = false;

	wlf_log(WLF_INFO, "Created macOS backend");
	return &backend->base;
}

static bool macos_backend_is_available(void) {
	// macOS backend is available if we're running on macOS
#ifdef __APPLE__
	return true;
#else
	return false;
#endif
}

static struct wlf_backend_registry_entry entry = {
	.type = WLF_BACKEND_MACOS,
	.name = "macos",
	.priority = 100,
	.create = macos_backend_create,
	.is_available = macos_backend_is_available,
	.handle = NULL,
};

bool wlf_backend_macos_register(void) {
	return wlf_backend_register(&entry);
}

bool wlf_backend_is_macos(struct wlf_backend *backend) {
	return (backend && backend->impl == &macos_backend_impl &&
			backend->type == WLF_BACKEND_MACOS);
}

struct wlf_backend_macos *wlf_backend_macos_from_backend(struct wlf_backend *wlf_backend) {
	assert(wlf_backend && wlf_backend->impl == &macos_backend_impl);
	struct wlf_backend_macos *backend = wlf_container_of(wlf_backend, backend, base);

	return backend;
}

File: platform/macos/backend.m
#include "wlf/platform/macos/backend.h"
#include "wlf/platform/wlf_backend.h"
#include "wlf/utils/wlf_log.h"
#include "wlf/utils/wlf_utils.h"

#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>

#include <assert.h>
#include <errno.h>
#include <poll.h>
#include <stdlib.h>

enum {
	WLF_MACOS_EVENT_LOOP_TIMEOUT_MS = 16,
};

static bool backend_ensure_appkit(struct wlf_backend_macos *macos) {
	WLF_UNUSED(macos);

	@autoreleasepool {
		if (NSApp == nil) {
			[NSApplication sharedApplication];
		}

		[NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
	}

	return true;
}

static void backend_dispatch_event_sources(struct wlf_backend *backend,
		int timeout_ms) {
	if (backend->event_source_count == 0) {
		if (timeout_ms > 0) {
			(void)poll(NULL, 0, timeout_ms);
		}
		return;
	}

	struct pollfd *fds = calloc(backend->event_source_count, sizeof(*fds));
	if (fds == NULL) {
		wlf_log_errno(WLF_ERROR, "Failed to allocate pollfd array");
		backend->running = false;
		return;
	}

	for (size_t i = 0; i < backend->event_source_count; ++i) {
		fds[i].fd = backend->event_sources[i].fd;
		fds[i].events = backend->event_sources[i].events;
	}

	int poll_ret = poll(fds, backend->event_source_count, timeout_ms);
	if (poll_ret < 0) {
		if (errno != EINTR) {
			wlf_log_errno(WLF_ERROR, "poll() failed in macOS event loop");
			backend->running = false;
		}
		free(fds);
		return;
	}

	if (poll_ret == 0) {
		free(fds);
		return;
	}

	for (size_t i = 0; i < backend->event_source_count; ++i) {
		if (fds[i].revents == 0 || backend->event_sources[i].dispatch == NULL) {
			continue;
		}

		backend->event_sources[i].dispatch(
			backend,
			backend->event_sources[i].fd,
			(uint32_t)fds[i].revents,
			backend->event_sources[i].data);
	}

	free(fds);
}

static bool backend_process_appkit_events(struct wlf_backend_macos *macos) {
	bool handled_event = false;

	@autoreleasepool {
		if (!backend_ensure_appkit(macos)) {
			return false;
		}

		for (;;) {
			NSEvent *event = [NSApp nextEventMatchingMask:NSEventMaskAny
				untilDate:[NSDate distantPast]
				inMode:NSDefaultRunLoopMode
				dequeue:YES];
			if (event == nil) {
				break;
			}

			handled_event = true;
			[NSApp sendEvent:event];
			[NSApp updateWindows];
		}
	}

	return handled_event;
}

static void *backend_native_display(struct wlf_backend *backend) {
	struct wlf_backend_macos *macos = wlf_backend_macos_from_backend(backend);
	if (!backend_ensure_appkit(macos)) {
		return NULL;
	}

	return (__bridge void *)NSApp;
}

static void backend_destroy(struct wlf_backend *backend) {
	struct wlf_backend_macos *macos = wlf_backend_macos_from_backend(backend);

	free(macos);
}

static void backend_exe(struct wlf_backend *backend) {
	struct wlf_backend_macos *macos = wlf_backend_macos_from_backend(backend);

	if (!backend_ensure_appkit(macos)) {
		backend->running = false;
		return;
	}

	backend->running = true;

	while (backend->running) {
		bool handled_appkit_event = backend_process_appkit_events(macos);

		/*
		 * Pump AppKit opportunistically, then block briefly on registered
		 * descriptors so external sources can participate in the backend loop.
		 */
		int timeout_ms = handled_appkit_event ? 0 : WLF_MACOS_EVENT_LOOP_TIMEOUT_MS;
		backend_dispatch_event_sources(backend, timeout_ms);
	}
}

static const struct wlf_backend_impl macos_backend_impl = {
	.name = "macOS",
	.destroy = backend_destroy,
	.exe = backend_exe,
	.native_display = backend_native_display,
};

struct wlf_backend *macos_backend_create(void) {
	struct wlf_backend_macos *backend = calloc(1, sizeof(struct wlf_backend_macos));
	if (backend == NULL) {
		wlf_log_errno(WLF_ERROR, "Failed to allocate wlf_backend_macos");
		return NULL;
	}

	wlf_backend_init(&backend->base, &macos_backend_impl);
	backend->base.data = backend;

	if (!backend_ensure_appkit(backend)) {
		free(backend);
		return NULL;
	}

	return &backend->base;
}

bool wlf_backend_macos_register(void) {
	wlf_log(WLF_DEBUG, "macOS backend registration is a no-op");
	return true;
}

bool wlf_backend_is_macos(struct wlf_backend *backend) {
	return backend != NULL && backend->impl == &macos_backend_impl;
}

struct wlf_backend_macos *wlf_backend_macos_from_backend(struct wlf_backend *backend) {
	struct wlf_backend_macos *macos = NULL;

	assert(backend != NULL);
	assert(backend->impl == &macos_backend_impl);

	return wlf_container_of(backend, macos, base);
}
